v2/metrics/datadog_test.go

```diff
@@ -0,0 +1,187 @@
+	assert.Equal(t, [][]float64{{mockClock, 123}}, p95.GetPoints())
+
+	p99 := p.Series[3]
+	assert.Equal(t, "voucher.check.run.latency.99percentile", p99.Metric)
+	assert.Equal(t, "gauge", p99.GetType())
+	assert.Equal(t, [][]float64{{mockClock, 123}}, p99.GetPoints())
+
+	max := p.Series[4]
+	assert.Equal(t, "voucher.check.run.latency.max", max.Metric)
+	assert.Equal(t, "gauge", max.GetType())
+	assert.Equal(t, [][]float64{{mockClock, 123}}, max.GetPoints())
+}
+
+func TestDatadogClient_Timing_Aggregate(t *testing.T) {
+	var p datadog.MetricsPayload
+	metrics, _ := newMockedDatadogClient(t, &p)
+
+	for i := 0; i < 100; i++ {
+		metrics.CheckRunLatency("diy", time.Duration(i+1)*time.Millisecond)
+	}
+	metrics.Close()
+
+	require.Len(t, p.Series, 5)
+	count := p.Series[0]
+	assert.Equal(t, "voucher.check.run.latency.count", count.Metric)
+	assert.Equal(t, "gauge", count.GetType())
+	assert.Equal(t, [][]float64{{mockClock, 100}}, count.GetPoints())
+
+	avg := p.Series[1]
+	assert.Equal(t, "voucher.check.run.latency.avg", avg.Metric)
+	assert.Equal(t, "gauge", avg.GetType())
+	assert.Equal(t, [][]float64{{mockClock, 50.5}}, avg.GetPoints())
+
+	p95 := p.Series[2]
+	assert.Equal(t, "voucher.check.run.latency.95percentile", p95.Metric)
+	assert.Equal(t, "gauge", p95.GetType())
+	assert.Equal(t, [][]float64{{mockClock, 95}}, p95.GetPoints())
+
+	p99 := p.Series[3]
+	assert.Equal(t, "voucher.check.run.latency.99percentile", p99.Metric)
+	assert.Equal(t, "gauge", p99.GetType())
+	assert.Equal(t, [][]float64{{mockClock, 99}}, p99.GetPoints())
+
+	max := p.Series[4]
+	assert.Equal(t, "voucher.check.run.latency.max", max.Metric)
+	assert.Equal(t, "gauge", max.GetType())
+	assert.Equal(t, [][]float64{{mockClock, 100}}, max.GetPoints())
+}
+
+func TestDatadogClient_AsyncFlush(t *testing.T) {
+	var p datadog.MetricsPayload
+	metrics, signal := newMockedDatadogClient(t, &p)
+
+	metrics.CheckRunStart("diy")
+
+	select {
+	case <-signal:
+	case <-time.After(2 * testSubmitInterval):
+		require.Fail(t, "expected flush did not happen")
+	}
+	require.Len(t, p.Series, 1)
+	series := p.Series[0]
+	assert.Equal(t, "voucher.check.run.start", series.Metric)
+	assert.Equal(t, []string{"check:diy"}, series.GetTags())
+	assert.Equal(t, "count", series.GetType())
+	assert.Equal(t, [][]float64{{mockClock, 1}}, series.GetPoints())
+}
+
+func TestDatadogClient_Event(t *testing.T) {
+	var p datadog.EventCreateRequest
+	metrics, _ := newMockedDatadogClient(t, &p)
+
+	metrics.CheckAttestationError("diy", errors.New("kaboom"))
+
+	assert.Equal(t, "kaboom", p.Text)
+	assert.Equal(t, "Voucher Check Attestation Error", p.Title)
+	assert.Equal(t, datadog.EVENTALERTTYPE_ERROR, p.GetAlertType())
+	assert.Equal(t, datadog.EVENTPRIORITY_LOW, p.GetPriority())
+	assert.Equal(t, []string{"check:diy"}, p.GetTags())
+	assert.Equal(t, int64(mockClock), p.GetDateHappened())
+}
+
+func newMockedDatadogClient(t testing.TB, data interface{}) (*metrics.DatadogClient, <-chan struct{}) {
```

Comment:
interesting question, how does the data get embedded within the datadog client? Related to question above

Reply:
For most tests, it's a pointer to a  [`datadog.MetricsPayload`](https://github.com/Shopify/voucher/blob/4eefd821466e5e547fe87da4ebb54719a4224510/v2/metrics/datadog_test.go#L127-L128). The exception is `TestDatadogClient_Event`, where it's a pointer to a [`datadog.EventCreateRequest`](https://github.com/Shopify/voucher/blob/4eefd821466e5e547fe87da4ebb54719a4224510/v2/metrics/datadog_test.go#L146-L147).
Whatever struct it is, the mock server [json deserializes](https://github.com/Shopify/voucher/blob/4eefd821466e5e547fe87da4ebb54719a4224510/v2/metrics/datadog_test.go#L167) whatever request it received onto that argument, before returning its response to the client. That means once our client under test has sent its POST to "datadog", the struct we passed to the mock server has captured the request.

That allows the test to be nice and type-safe: it makes assertions against a `datadog.MetricsPayload` instead of a JSON-ified `string`, or the dreaded `map[string]interface{}`.